database/memory/db.go

```diff
@@ -535,34 +611,199 @@ func (db *MemoryDB) removeAllIndices(address types.Address) error {
 }
 
 func (db *MemoryDB) GetAllTokenHolders(contract types.Address, block uint64, options *types.TokenQueryOptions) ([]types.Address, error) {
-	return nil, database.ErrNotImplemented
+	var holderMap = make(map[types.Address]bool)
+	for _, k := range db.erc20BalancesDB {
+		if k.Contract == contract && k.BlockNumber <= block && k.Holder != "0000000000000000000000000000000000000000" {
+			holderMap[k.Holder] = true
+		}
+	}
+	var holderArr []types.Address
+	for holdr := range holderMap {
+		holderArr = append(holderArr, holdr)
+	}
+	return holderArr, nil
 }
 
 func (db *MemoryDB) RecordERC721Token(contract types.Address, holder types.Address, block uint64, tokenId *big.Int) error {
+	//find old entry
+	existingTokenEntry, errExisting := db.ERC721TokenByTokenID(contract, block-1, tokenId)
+	if errExisting != nil && errExisting != database.ErrNotFound {
+		return errExisting
+	}
+
+	paddedTokenId := fmt.Sprintf("%085d", tokenId)
+	first, _ := strconv.ParseUint(paddedTokenId[0:17], 10, 64)
+	second, _ := strconv.ParseUint(paddedTokenId[17:34], 10, 64)
+	third, _ := strconv.ParseUint(paddedTokenId[34:51], 10, 64)
+	fourth, _ := strconv.ParseUint(paddedTokenId[51:68], 10, 64)
+	fifth, _ := strconv.ParseUint(paddedTokenId[68:85], 10, 64)
+
+	//add new entry
+	tokenHolderInfo := SortableERC721Token{
+		types.ERC721Token{
+			Contract:  contract,
+			Holder:    holder,
+			Token:     tokenId.String(),
+			HeldFrom:  block,
+			HeldUntil: nil,
+		},
+		first, second, third, fourth, fifth,
+	}
+	db.erc721BalancesDB = append(db.erc721BalancesDB, tokenHolderInfo)
+	/////
+	if errExisting == database.ErrNotFound {
+		return nil
+	}
+
+	blk := block - 1
+	existingTokenEntry.HeldUntil = &blk
 	return nil
 }
 
-func (db *MemoryDB) ERC721TokenByTokenID(contract types.Address, block uint64, tokenId *big.Int) (types.ERC721Token, error) {
-	return types.ERC721Token{}, database.ErrNotImplemented
+func (db *MemoryDB) ERC721TokenByTokenID(contract types.Address, block uint64, tokenId *big.Int) (*types.ERC721Token, error) {
+	var tmpItem int
+	found := false
+	for i, item := range db.erc721BalancesDB {
+		if item.Contract == contract && item.HeldFrom <= block && item.Token == tokenId.String() {
+			if !found {
+				tmpItem = i
+				found = true
+			} else {
+				if item.HeldFrom > db.erc721BalancesDB[tmpItem].HeldFrom {
+					tmpItem = i
+				}
+			}
+
+		}
+	}
+	if !found {
+		return nil, database.ErrNotFound
+	}
+	return &db.erc721BalancesDB[tmpItem].ERC721Token, nil
 }
 
 func (db *MemoryDB) ERC721TokensForAccountAtBlock(contract types.Address, holder types.Address, block uint64, options *types.TokenQueryOptions) ([]types.ERC721Token, error) {
-	return nil, database.ErrNotImplemented
+	return db.erc721TokensAtBlock(contract, &holder, block, options)
+}
+
+func (db *MemoryDB) erc721TokensAtBlock(contract types.Address, holder *types.Address, block uint64, options *types.TokenQueryOptions) ([]types.ERC721Token, error) {
+	startTokenId := big.NewInt(-1)
```

Comment:
please add a read lock to avoid concurrency

Reply:
done